File: swift-bridge/Sources/ModelIOBridge/VertexAttribute.swift
import Foundation
import ModelIO
import simd

private func mdl_vertex_attribute_info(_ attribute: MDLVertexAttribute) -> [String: Any] {
    [
        "name": attribute.name,
        "format": attribute.format.rawValue,
        "offset": attribute.offset,
        "buffer_index": attribute.bufferIndex,
        "time": attribute.time,
        "initialization_value": [attribute.initializationValue.x, attribute.initializationValue.y, attribute.initializationValue.z, attribute.initializationValue.w],
    ]
}

private func mdl_vertex_descriptor_info(_ descriptor: MDLVertexDescriptor) -> [String: Any] {
    let attributes = descriptor.attributes.compactMap { $0 as? MDLVertexAttribute }.map(mdl_vertex_attribute_info)
    let layoutStrides = descriptor.layouts.compactMap { ($0 as? MDLVertexBufferLayout)?.stride }
    return [
        "attribute_count": attributes.count,
        "attributes": attributes,
        "layout_strides": layoutStrides,
    ]
}

@_cdecl("mdl_vertex_attribute_new")
public func mdl_vertex_attribute_new(
    _ name: UnsafePointer<CChar>?,
    _ formatRaw: UInt32,
    _ offset: UInt64,
    _ bufferIndex: UInt64,
    _ outAttribute: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    mdl_run(outError) {
        guard let name, let outAttribute else {
            throw ModelIOBridgeError.invalidArgument("missing vertex attribute name or output pointer")
        }
        outAttribute.pointee = mdl_retain(
            MDLVertexAttribute(
                name: String(cString: name),
                format: try mdl_vertex_format(formatRaw),
                offset: Int(offset),
                bufferIndex: Int(bufferIndex)
            )
        )
    }
}

@_cdecl("mdl_vertex_attribute_info_json")
public func mdl_vertex_attribute_info_json(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let attribute = mdl_borrow_object(handle) as? MDLVertexAttribute else { return nil }
    return mdl_string(mdl_json_string(from: mdl_vertex_attribute_info(attribute)) ?? "{}")
}

@_cdecl("mdl_vertex_attribute_set_name")
public func mdl_vertex_attribute_set_name(_ handle: UnsafeMutableRawPointer?, _ name: UnsafePointer<CChar>?) {
    guard let attribute = mdl_borrow_object(handle) as? MDLVertexAttribute,
          let name
    else {
        return
    }
    attribute.name = String(cString: name)
}

@_cdecl("mdl_vertex_attribute_set_format")
public func mdl_vertex_attribute_set_format(_ handle: UnsafeMutableRawPointer?, _ rawValue: UInt32) {
    guard let attribute = mdl_borrow_object(handle) as? MDLVertexAttribute,
          let format = try? mdl_vertex_format(rawValue)
    else {
        return
    }
    attribute.format = format
}

@_cdecl("mdl_vertex_attribute_set_offset")
public func mdl_vertex_attribute_set_offset(_ handle: UnsafeMutableRawPointer?, _ offset: UInt64) {
    guard let attribute = mdl_borrow_object(handle) as? MDLVertexAttribute else { return }
    attribute.offset = Int(offset)
}

@_cdecl("mdl_vertex_attribute_set_buffer_index")
public func mdl_vertex_attribute_set_buffer_index(_ handle: UnsafeMutableRawPointer?, _ bufferIndex: UInt64) {
    guard let attribute = mdl_borrow_object(handle) as? MDLVertexAttribute else { return }
    attribute.bufferIndex = Int(bufferIndex)
}

@_cdecl("mdl_vertex_attribute_set_time")
public func mdl_vertex_attribute_set_time(_ handle: UnsafeMutableRawPointer?, _ time: Double) {
    guard let attribute = mdl_borrow_object(handle) as? MDLVertexAttribute else { return }
    attribute.time = time
}

@_cdecl("mdl_vertex_attribute_set_initialization_value")
public func mdl_vertex_attribute_set_initialization_value(_ handle: UnsafeMutableRawPointer?, _ x: Float, _ y: Float, _ z: Float, _ w: Float) {
    guard let attribute = mdl_borrow_object(handle) as? MDLVertexAttribute else { return }
    attribute.initializationValue = SIMD4<Float>(x, y, z, w)
}

@_cdecl("mdl_vertex_descriptor_new_copy")
public func mdl_vertex_descriptor_new_copy(
    _ handle: UnsafeMutableRawPointer?,
    _ outDescriptor: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    mdl_run(outError) {
        guard let descriptor = mdl_borrow_object(handle) as? MDLVertexDescriptor,
              let outDescriptor
        else {
            throw ModelIOBridgeError.invalidArgument("missing source vertex descriptor or output pointer")
        }
        outDescriptor.pointee = mdl_retain(MDLVertexDescriptor(vertexDescriptor: descriptor))
    }
}

@_cdecl("mdl_vertex_descriptor_info_json")
public func mdl_vertex_descriptor_info_json(_ handle: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let descriptor = mdl_borrow_object(handle) as? MDLVertexDescriptor else { return nil }
    return mdl_string(mdl_json_string(from: mdl_vertex_descriptor_info(descriptor)) ?? "{}")
}

@_cdecl("mdl_vertex_descriptor_attribute_count")
public func mdl_vertex_descriptor_attribute_count(_ handle: UnsafeMutableRawPointer?) -> UInt64 {
    guard let descriptor = mdl_borrow_object(handle) as? MDLVertexDescriptor else { return 0 }
    return UInt64(descriptor.attributes.count)
}

@_cdecl("mdl_vertex_descriptor_attribute_at")
public func mdl_vertex_descriptor_attribute_at(_ handle: UnsafeMutableRawPointer?, _ index: UInt64) -> UnsafeMutableRawPointer? {
    guard let descriptor = mdl_borrow_object(handle) as? MDLVertexDescriptor,
          index < UInt64(descriptor.attributes.count),
          let attribute = descriptor.attributes[Int(index)] as? MDLVertexAttribute
    else {
        return nil
    }
    return mdl_retain(attribute)
}

@_cdecl("mdl_vertex_descriptor_attribute_named")
public func mdl_vertex_descriptor_attribute_named(_ handle: UnsafeMutableRawPointer?, _ name: UnsafePointer<CChar>?) -> UnsafeMutableRawPointer? {
    guard let descriptor = mdl_borrow_object(handle) as? MDLVertexDescriptor,
          let name,
          let attribute = descriptor.attributeNamed(String(cString: name))
    else {
        return nil
    }
    return mdl_retain(attribute)
}

@_cdecl("mdl_vertex_descriptor_reset")
public func mdl_vertex_descriptor_reset(_ handle: UnsafeMutableRawPointer?) {
    guard let descriptor = mdl_borrow_object(handle) as? MDLVertexDescriptor else { return }
    descriptor.reset()
}

@_cdecl("mdl_vertex_descriptor_set_packed_offsets")
public func mdl_vertex_descriptor_set_packed_offsets(_ handle: UnsafeMutableRawPointer?) {
    guard let descriptor = mdl_borrow_object(handle) as? MDLVertexDescriptor else { return }
    descriptor.setPackedOffsets()
}

@_cdecl("mdl_vertex_descriptor_set_packed_strides")
public func mdl_vertex_descriptor_set_packed_strides(_ handle: UnsafeMutableRawPointer?) {
    guard let descriptor = mdl_borrow_object(handle) as? MDLVertexDescriptor else { return }
    descriptor.setPackedStrides()
}

File: src/vertex_attribute.rs
use std::ptr;

use crate::error::Result;
use crate::ffi;
use crate::handle::ObjectHandle;
use crate::types::{VertexAttributeDescriptorInfo, VertexDescriptorInfo};
use crate::util::{c_string, parse_json, required_handle};

#[derive(Debug, Clone)]
/// Wraps the corresponding Model I/O vertex attribute counterpart.
pub struct VertexAttribute {
    handle: ObjectHandle,
}

impl VertexAttribute {
    /// Builds this wrapper from the retained handle of the wrapped Model I/O vertex attribute counterpart.
    pub(crate) fn from_handle(handle: ObjectHandle) -> Self {
        Self { handle }
    }

    /// Wraps the corresponding Model I/O initializer for the wrapped Model I/O vertex attribute counterpart.
    pub fn new(name: &str, format: u32, offset: usize, buffer_index: usize) -> Result<Self> {
        let name = c_string(name)?;
        let mut out_attribute = ptr::null_mut();
        let mut out_error = ptr::null_mut();
        // SAFETY: The unsafe operation is valid in this context.
        let status = unsafe {
            ffi::mdl_vertex_attribute_new(
                name.as_ptr(),
                format,
                offset as u64,
                buffer_index as u64,
                &mut out_attribute,
                &mut out_error,
            )
        };
        crate::util::status_result(status, out_error)?;
        Ok(Self::from_handle(required_handle(
            out_attribute,
            "MDLVertexAttribute",
        )?))
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex attribute counterpart.
    pub fn info(&self) -> Result<VertexAttributeDescriptorInfo> {
        parse_json(
            // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
            unsafe { ffi::mdl_vertex_attribute_info_json(self.handle.as_ptr()) },
            "MDLVertexAttribute",
        )
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex attribute counterpart.
    pub fn set_name(&self, name: &str) -> Result<()> {
        let name = c_string(name)?;
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_vertex_attribute_set_name(self.handle.as_ptr(), name.as_ptr()) };
        Ok(())
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex attribute counterpart.
    pub fn set_format(&self, format: u32) {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_vertex_attribute_set_format(self.handle.as_ptr(), format) };
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex attribute counterpart.
    pub fn set_offset(&self, offset: usize) {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_vertex_attribute_set_offset(self.handle.as_ptr(), offset as u64) };
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex attribute counterpart.
    pub fn set_buffer_index(&self, buffer_index: usize) {
        // SAFETY: The unsafe operation is valid in this context.
        unsafe {
            ffi::mdl_vertex_attribute_set_buffer_index(self.handle.as_ptr(), buffer_index as u64);
        };
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex attribute counterpart.
    pub fn set_time(&self, time: f64) {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_vertex_attribute_set_time(self.handle.as_ptr(), time) };
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex attribute counterpart.
    pub fn set_initialization_value(&self, value: [f32; 4]) {
        // SAFETY: The unsafe operation is valid in this context.
        unsafe {
            ffi::mdl_vertex_attribute_set_initialization_value(
                self.handle.as_ptr(),
                value[0],
                value[1],
                value[2],
                value[3],
            );
        };
    }
}

#[derive(Debug, Clone)]
/// Wraps the corresponding Model I/O vertex buffer layout counterpart.
pub struct VertexBufferLayout {
    handle: ObjectHandle,
}

impl VertexBufferLayout {
    /// Builds this wrapper from the retained handle of the wrapped Model I/O vertex buffer layout counterpart.
    pub(crate) fn from_handle(handle: ObjectHandle) -> Self {
        Self { handle }
    }

    /// Wraps the corresponding Model I/O initializer for the wrapped Model I/O vertex buffer layout counterpart.
    pub fn new(stride: usize) -> Result<Self> {
        let mut out_layout = ptr::null_mut();
        let mut out_error = ptr::null_mut();
        // SAFETY: The unsafe operation is valid in this context.
        let status = unsafe {
            ffi::mdl_vertex_buffer_layout_new(stride as u64, &mut out_layout, &mut out_error)
        };
        crate::util::status_result(status, out_error)?;
        Ok(Self::from_handle(required_handle(
            out_layout,
            "MDLVertexBufferLayout",
        )?))
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex buffer layout counterpart.
    pub fn stride(&self) -> usize {
        // SAFETY: Output pointers are initialized and managed; FFI function is called safely.
        unsafe { ffi::mdl_vertex_buffer_layout_stride(self.handle.as_ptr()) as usize }
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex buffer layout counterpart.
    pub fn set_stride(&self, stride: usize) {
        // SAFETY: Output pointers are initialized and managed; FFI function is called safely.
        unsafe { ffi::mdl_vertex_buffer_layout_set_stride(self.handle.as_ptr(), stride as u64) };
    }
}

#[derive(Debug, Clone)]
/// Wraps the corresponding Model I/O vertex descriptor counterpart.
pub struct VertexDescriptor {
    handle: ObjectHandle,
}

impl VertexDescriptor {
    /// Builds this wrapper from the retained handle of the wrapped Model I/O vertex descriptor counterpart.
    pub(crate) fn from_handle(handle: ObjectHandle) -> Self {
        Self { handle }
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex descriptor counterpart.
    pub fn copy(&self) -> Result<Self> {
        let mut out_descriptor = ptr::null_mut();
        let mut out_error = ptr::null_mut();
        // SAFETY: The unsafe operation is valid in this context.
        let status = unsafe {
            ffi::mdl_vertex_descriptor_new_copy(
                self.handle.as_ptr(),
                &mut out_descriptor,
                &mut out_error,
            )
        };
        crate::util::status_result(status, out_error)?;
        Ok(Self::from_handle(required_handle(
            out_descriptor,
            "MDLVertexDescriptor",
        )?))
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex descriptor counterpart.
    pub fn info(&self) -> Result<VertexDescriptorInfo> {
        parse_json(
            // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
            unsafe { ffi::mdl_vertex_descriptor_info_json(self.handle.as_ptr()) },
            "MDLVertexDescriptor",
        )
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex descriptor counterpart.
    pub fn attribute_count(&self) -> usize {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_vertex_descriptor_attribute_count(self.handle.as_ptr()) as usize }
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex descriptor counterpart.
    pub fn attribute_at(&self, index: usize) -> Option<VertexAttribute> {
        let ptr =
            // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
            unsafe { ffi::mdl_vertex_descriptor_attribute_at(self.handle.as_ptr(), index as u64) };
        // SAFETY: The unsafe operation is valid in this context.
        unsafe { ObjectHandle::from_retained_ptr(ptr) }.map(VertexAttribute::from_handle)
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex descriptor counterpart.
    pub fn attribute_named(&self, name: &str) -> Result<Option<VertexAttribute>> {
        let name = c_string(name)?;
        // SAFETY: The unsafe operation is valid in this context.
        let ptr = unsafe {
            ffi::mdl_vertex_descriptor_attribute_named(self.handle.as_ptr(), name.as_ptr())
        };
        // SAFETY: The unsafe operation is valid in this context.
        Ok(unsafe { ObjectHandle::from_retained_ptr(ptr) }.map(VertexAttribute::from_handle))
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex descriptor counterpart.
    pub fn attributes(&self) -> Vec<VertexAttribute> {
        (0..self.attribute_count())
            .filter_map(|index| self.attribute_at(index))
            .collect()
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex descriptor counterpart.
    pub fn layout_count(&self) -> usize {
        // SAFETY: Output pointers are initialized and managed; FFI function is called safely.
        unsafe { ffi::mdl_vertex_descriptor_layout_count(self.handle.as_ptr()) as usize }
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex descriptor counterpart.
    pub fn layout_at(&self, index: usize) -> Option<VertexBufferLayout> {
        let ptr =
            // SAFETY: Output pointers are initialized and managed; FFI function is called safely.
            unsafe { ffi::mdl_vertex_descriptor_layout_at(self.handle.as_ptr(), index as u64) };
        // SAFETY: The unsafe operation is valid in this context.
        unsafe { ObjectHandle::from_retained_ptr(ptr) }.map(VertexBufferLayout::from_handle)
    }

    #[must_use]
    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex descriptor counterpart.
    pub fn layouts(&self) -> Vec<VertexBufferLayout> {
        (0..self.layout_count())
            .filter_map(|index| self.layout_at(index))
            .collect()
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex descriptor counterpart.
    pub fn reset(&self) {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_vertex_descriptor_reset(self.handle.as_ptr()) };
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex descriptor counterpart.
    pub fn set_packed_offsets(&self) {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_vertex_descriptor_set_packed_offsets(self.handle.as_ptr()) };
    }

    /// Calls the corresponding Model I/O method on the wrapped Model I/O vertex descriptor counterpart.
    pub fn set_packed_strides(&self) {
        // SAFETY: ObjectHandle wraps a valid opaque pointer from Swift; FFI function accepts it safely.
        unsafe { ffi::mdl_vertex_descriptor_set_packed_strides(self.handle.as_ptr()) };
    }
}
